src/main/java/com/google/sps/servlets/EventsServlet.java

```diff
@@ -0,0 +1,199 @@
+    static final String DATE = "date";
+    static final String START_TIME = "start";
+    static final String END_TIME = "end";
+    static final String USER_TIMEZONE = "timezone";
+
+    static final List<String> CATEGORIES = new ArrayList<String>(
+        Arrays.asList("food_beverage", "nature", "water", "waste_cleanup", "other")
+    );
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    FetchOptions fetchOptions = FetchOptions.Builder.withLimit(10);
+    Query query = new Query(EVENT).addSort(TIMESTAMP, SortDirection.DESCENDING);
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery pq = datastore.prepare(query);
+
+    QueryResultList<Entity> results = pq.asQueryResultList(fetchOptions);
+
+    List<Event> events = new ArrayList<>();
+    for (Entity entity : results) {
+        long timestamp = (long) entity.getProperty(TIMESTAMP);
+        String summary = (String) entity.getProperty(SUMMARY);
+        String description = (String) entity.getProperty(DESCRIPTION);
+        String location = (String) entity.getProperty(LOCATION);
+        Date startTime = (Date) entity.getProperty(START_TIME);
+        Date endTime = (Date) entity.getProperty(END_TIME);
+        String category = (String) entity.getProperty(CATEGORY);
+
+        DateTime startDateTime = new DateTime(startTime);
+        EventDateTime start = new EventDateTime()
+            .setDateTime(startDateTime)
+            .setTimeZone(UTC_TIMEZONE);
+        DateTime endDateTime = new DateTime(endTime);
+        EventDateTime end = new EventDateTime()
+            .setDateTime(endDateTime)
+            .setTimeZone(UTC_TIMEZONE);
+        Event event = new Event()
+            .setSummary(summary)
+            .setLocation(location)
+            .setDescription(description)
+            .setStart(start)
+            .setEnd(end);
+
+        ExtendedProperties ep = new ExtendedProperties();
+        ep.set(CATEGORY, category);
+        event.setExtendedProperties(ep);
+    
+        events.add(event);
+    }
+
+    response.setContentType("application/json; charset=UTF-8");
+    response.setCharacterEncoding("UTF-8");
+
+    String json = convertToJson(events);
+    response.getWriter().println(json);
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+      String eventSummary = request.getParameter(SUMMARY);
+      String eventDescription = request.getParameter(DESCRIPTION);
+      String eventLocation = request.getParameter(LOCATION);
+      String eventDateString = request.getParameter(DATE);
+      String eventStartTimeString = request.getParameter(START_TIME);
+      String eventEndTimeString = request.getParameter(END_TIME);
+      String timezoneOffset = request.getParameter(USER_TIMEZONE);
+      String category = request.getParameter(CATEGORY);
```

Comment:
You may want to sanitize all user input here.

E.g. Datastore throws an error for strings > 1500 bytes (http://cloud/appengine/docs/standard/java/datastore/entities#Properties_and_value_types), so you might want to truncate extra long summary, description, location strings. Moreover, you can also ensure strings are UTF-8 compliant https://stackoverflow.com/q/6622226.


Slightly off topic, but you may want to read the incredibly useful article on Unicode. It's been a lifesaver for me so many times: https://www.joelonsoftware.com/2003/10/08/the-absolute-minimum-every-software-developer-absolutely-positively-must-know-about-unicode-and-character-sets-no-excuses/